src/sources/exec.rs

```diff
@@ -0,0 +1,868 @@
+                    Ok(output) => {
+                        let mut events = Vec::new();
+                        let elapsed_millis = now.elapsed().as_millis();
+                        let exit_status = output.status.code();
+
+                        let stdout_events = process_scheduled_output(
+                            config,
+                            event_per_line,
+                            exec_duration_millis_key.clone(),
+                            pid,
+                            exit_status,
+                            elapsed_millis,
+                            output.stdout,
+                            STDOUT,
+                        );
+
+                        if let Some(mut stdout_events) = stdout_events {
+                            events.append(&mut stdout_events);
+                        }
+
+                        // Optionally include stderr
+                        if config.include_stderr.unwrap_or(true) {
+                            let stderr_events = process_scheduled_output(
+                                config,
+                                event_per_line,
+                                exec_duration_millis_key.clone(),
+                                pid,
+                                exit_status,
+                                elapsed_millis,
+                                output.stderr,
+                                STDERR,
+                            );
+
+                            if let Some(mut stderr_events) = stderr_events {
+                                events.append(&mut stderr_events);
+                            }
+                        }
+
+                        if events.is_empty() {
+                            None
+                        } else {
+                            Some(events)
+                        }
+                    }
+                    Err(command_error) => {
+                        error!("Error during scheduled exec run.");
+                        emit!(ExecFailed {
+                            error: command_error
+                        });
+                        None
+                    }
+                },
+                Err(_) => {
+                    error!("Timed out during scheduled exec run.");
+                    emit!(ExecTimeout {
+                        elapsed_millis: now.elapsed().as_secs()
+                    });
+                    None
+                }
+            }
+        }
+        Err(spawn_error) => {
+            error!("Error during scheduled exec run.");
+            emit!(ExecFailed { error: spawn_error });
+            None
+        }
+    }
+}
+
+fn process_scheduled_output(
+    config: &ExecConfig,
+    event_per_line: bool,
+    exec_duration_millis_key: String,
+    pid: u32,
+    exit_status: Option<i32>,
+    elapsed_millis: u128,
+    output: Vec<u8>,
+    stream: &str,
+) -> Option<Vec<Event>> {
+    let output_string: String = String::from_utf8(output).unwrap();
```

Comment:
We just treat these as bytes internally so I don't think you need to decode as UTF8 here just to convert to `Bytes` again later.

Reply:
The logic is different for the streaming processing (which we do for both now). There is conversion to utf8 as part of that processing but I think that is useful to ensure we are producing a valid utf8 string.